samza-core/src/main/java/org/apache/samza/operators/StreamGraphSpec.java

```diff
@@ -69,87 +77,98 @@
    */
   private int nextOpNum = 0;
   private final Set<String> operatorIds = new HashSet<>();
-  private Serde<?> defaultSerde = new KVSerde(new NoOpSerde(), new NoOpSerde());
   private ContextManager contextManager = null;
+  private Optional<SystemDescriptor> defaultSystemDescriptorOptional = Optional.empty();
 
   public StreamGraphSpec(Config config) {
     this.config = config;
   }
 
   @Override
-  public void setDefaultSerde(Serde<?> serde) {
-    Preconditions.checkNotNull(serde, "Default serde must not be null");
+  public void setDefaultSystem(SystemDescriptor<?> defaultSystemDescriptor) {
+    Preconditions.checkNotNull(defaultSystemDescriptor, "Provided defaultSystemDescriptor must not be null.");
+    String defaultSystemName = defaultSystemDescriptor.getSystemName();
     Preconditions.checkState(inputOperators.isEmpty() && outputStreams.isEmpty(),
-        "Default serde must be set before creating any input or output streams.");
-    this.defaultSerde = serde;
+        "Default system must be set before creating any input or output streams.");
+    checkSystemDescriptorUniqueness(defaultSystemDescriptor, defaultSystemName);
+    systemDescriptors.put(defaultSystemName, defaultSystemDescriptor);
+    this.defaultSystemDescriptorOptional = Optional.of(defaultSystemDescriptor);
   }
 
   @Override
-  public <M> MessageStream<M> getInputStream(String streamId, Serde<M> serde) {
-    Preconditions.checkState(isValidId(streamId),
-        "streamId must be non-empty and must not contain spaces or special characters: " + streamId);
-    Preconditions.checkNotNull(serde, "serde must not be null for an input stream.");
+  public <M> MessageStream<M> getInputStream(InputDescriptor<M, ?> inputDescriptor) {
+    SystemDescriptor systemDescriptor = inputDescriptor.getSystemDescriptor();
+    Optional<StreamExpander> expander = systemDescriptor.getExpander();
+    if (expander.isPresent()) {
+      return expander.get().apply(this, inputDescriptor);
+    }
+
+    String streamId = inputDescriptor.getStreamId();
     Preconditions.checkState(!inputOperators.containsKey(streamId),
         "getInputStream must not be called multiple times with the same streamId: " + streamId);
+    Preconditions.checkState(!inputDescriptors.containsKey(streamId),
```

Comment:
So, in high-level API, if I call StreamGraphSpec.getInputDescriptors(), it will not include the intermediate streams? Should we explicitly create a list of intermediate streams in this case? Here is the discrepancy I saw: a) in ExecutionPlanner.createJobGraph(), the intermediate stream is calculated by intersection between input streams (from getInputOperators()) and output streams (from getOutputStreams()); b) in low-level application, there is no explicit notion of intermediate stream and if there is a need to calculate it, it will have to compute the intersection between InputDescriptors and OutputDescriptors as well. That would be different from the list of InputDescriptor maintained here. I would suggest make them consistent, either by enforcing  InputDescriptors have the same keySet with InputOperators, or explicitly maintain an intermediate streams list in both high and low level APIs to avoid the additional logic of computing intersections in the planner.

Reply:
That's right, currently the planner calculates the intersection of input and output streams to find intermediate streams for high level API. What would be the use case for supporting intermediate streams in low level API? I'm OK with maintaining a separate list of intermediate streams if we need them in low level API, but would prefer to do that as part of a separate PR (ideally, one that makes the Planner work for both low level and high level API jobs).